packages/wrangler/src/core/register-commands.ts

```diff
@@ -0,0 +1,236 @@
+import { COMMAND_DEFINITIONS } from "./define-command";
+import { wrapCommandDefinition } from "./wrap-command";
+import type { CommonYargsArgv, SubHelp } from "../yargs-types";
+import type {
+	AliasDefinition,
+	Command,
+	CommandDefinition,
+	NamespaceDefinition,
+} from "./define-command";
+
+export class CommandRegistrationError extends Error {}
+
+export function createCommandRegister(
+	yargs: CommonYargsArgv,
+	subHelp: SubHelp
+) {
+	const tree = createCommandTree();
+
+	return {
+		registerAll() {
+			for (const [segment, node] of tree.entries()) {
+				walkTreeAndRegister(segment, node, yargs, subHelp);
+				tree.delete(segment);
+			}
+		},
+		registerNamespace(namespace: string) {
+			const node = tree.get(namespace);
+
+			if (!node) {
+				throw new CommandRegistrationError(
+					`No definition found for namespace '${namespace}'`
+				);
+			}
+
+			tree.delete(namespace);
+			walkTreeAndRegister(namespace, node, yargs, subHelp);
+		},
+	};
+}
+
+type DefinitionTreeNode = {
+	definition?: CommandDefinition | NamespaceDefinition | AliasDefinition;
+	subtree: DefinitionTree;
+};
+type DefinitionTree = Map<string, DefinitionTreeNode>;
+
+type ResolvedDefinitionTreeNode = {
+	definition?: CommandDefinition | NamespaceDefinition;
+	subtree: ResolvedDefinitionTree;
+};
+type ResolvedDefinitionTree = Map<string, ResolvedDefinitionTreeNode>;
+
+/**
+ * Converts a flat list of COMMAND_DEFINITIONS into a tree of defintions
+ * which can be passed to yargs builder api
+ *
+ * For example,
+ *      wrangler dev
+ *      wrangler deploy
+ *      wrangler versions upload
+ *      wrangler versions deploy
+ *
+ * Will be transformed into:
+ *      wrangler
+ *           dev
+ *           deploy
+ *           versions
+ *               upload
+ *               deploy
+ */
+function createCommandTree() {
+	const root: DefinitionTreeNode = { subtree: new Map() };
+	const aliases = new Set<AliasDefinition>();
+
+	// STEP 1: Create tree from flat definitions array
+
+	for (const def of COMMAND_DEFINITIONS) {
+		const node = createNodeFor(def.command, root);
+
+		if (node.definition) {
```

Comment:
How will `node.definition` be defined if it's just been created with `createNodeFor()`?

Reply:
The node has been created, or returned if already exists. The definition is undefined it it was just created.